lib/Sema/CSApply.cpp

```diff
@@ -3656,6 +3656,27 @@ namespace {
             : CheckedCastContextKind::ConditionalCast;
 
       auto fromType = cs.getType(sub);
+
+      // Special handle for literals conditional checked cast when they can be
+      // statically coerced to the cast type.
+      if (isa<LiteralExpr>(sub)) {
+        if (auto protocol = TypeChecker::getLiteralProtocol(ctx, sub)) {
+          if (!fromType->isEqual(toType) &&
```

Comment:
What happens if the to type is more complex like NSString or NSNumber?

Reply:
This is still emitted, and the same is also valid for any type that conforms to the actual literal protocol can be coerced e.g. 
```swift
struct S: ExpressibleByStringLiteral, CustomStringConvertible {
  typealias StringLiteralType = String
  
  var _value: String = ""
  init(stringLiteral value: Self.StringLiteralType) {
    self._value = value
  }
  
  var description: String { _value }
}

let a = "A" as S
 
print(a) // "A"
```

I'm just adding the test cases :)

Reply:
Note: When the literal inferred type can be bridged to the cast type, the conditional downcast always succeeds, also added the test cases :)